src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,278 @@
+#include "controllers/rendering/controllerrenderingengine.h"
+
+#include <QEvent>
+#include <QLabel>
+#include <QOffscreenSurface>
+#include <QOpenGLBuffer>
+#include <QOpenGLContext>
+#include <QOpenGLFramebufferObject>
+#include <QOpenGLFunctions>
+#include <QOpenGLShaderProgram>
+#include <QOpenGLVertexArrayObject>
+#include <QQmlComponent>
+#include <QQmlEngine>
+#include <QQuickGraphicsDevice>
+#include <QQuickItem>
+#include <QQuickRenderControl>
+#include <QQuickRenderTarget>
+#include <QQuickWindow>
+#include <QScreen>
+
+#include "controllers/controller.h"
+#include "controllers/scripting/legacy/controllerscriptenginelegacy.h"
+#include "controllers/scripting/legacy/controllerscriptinterfacelegacy.h"
+#include "qml/qmlwaveformoverview.h"
+
+ControllerRenderingEngine::ControllerRenderingEngine(Controller* controller,
+        LegacyControllerMapping::QMLFileInfo qml,
+        const RuntimeLoggingCategory& logger,
+        uint8_t screenId)
+        : QObject(),
+          m_screenId(screenId),
+          m_renderingInfo(qml) {
+    m_pThread = std::make_unique<QThread>();
+    m_pThread->setObjectName("ControllerScreenRenderer");
+
+    if (!m_fileWatcher.addPath(qml.file.absoluteFilePath())) {
+        qCWarning(logger) << "Failed to watch render file" << qml.file.absoluteFilePath();
+    }
+
+    for (auto path : qml.libraries) {
+        // TODO make recursive?
+        QDir libDirectory(path.absoluteFilePath());
+        QFileInfoList fileEntries(libDirectory.entryInfoList(QStringList{"*.qml"}));
+        QStringList files;
+        for (auto entry : fileEntries) {
+            files << entry.absoluteFilePath();
+        }
+        files << path.absoluteFilePath();
+
+        QStringList failed(m_fileWatcher.addPaths(files));
+        if (!failed.isEmpty()) {
+            qCWarning(logger) << "Failed to watch render files" << failed
+                              << "in library path" << path.absoluteFilePath();
+        }
+    }
+
+    moveToThread(m_pThread.get());
+
+    connect(&m_fileWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &ControllerRenderingEngine::reload);
+
+    connect(m_pThread.get(), &QThread::started, this, &ControllerRenderingEngine::start);
+
+    m_pThread->start(QThread::NormalPriority);
+}
+
+void ControllerRenderingEngine::reload() {
+    m_fbo.reset();
+
+    m_renderControl.reset();
+    m_qmlComponent.reset();
+    m_quickWindow.reset();
+    m_qmlEngine.reset();
+
+    m_offscreenSurface.reset();
+
+    m_context.reset();
+
+    start();
+}
+
+void ControllerRenderingEngine::start() {
+    QSurfaceFormat format;
+    // Qt Quick may need a depth and stencil buffer. Always make sure these are available.
+    format.setDepthBufferSize(16);
+    format.setStencilBufferSize(8);
```

Comment:
```suggestion
    QSurfaceFormat format;
    // Qt Quick may need a depth and stencil buffer. Always make sure these are available.
    format.setDepthBufferSize(16);
    format.setStencilBufferSize(8);
    format.setRedBufferSize(5);
    format.setGreenBufferSize(6);
    format.setBlueBufferSize(5);
```
For direct rendering to RGB565

Reply:
Yep, I tried that already, but it doesn't seem to work. I'm suspecting the FBO to override the OpenGL format when calling `toImage()`, but I'm not quite sure how I could read the buffer without this helper function - perhaps I should mention that my OpenGL experience/knowledge/confidence is somewhere between zero and minus one...